### ISSUE TITLE
HTTP 503s when targeting a newly created ready backends

### ISSUE THREAD

**Issue by dprotaso** [istio/istio#43816]
### Bug Description

Knative has a networking conformance test were we probe an ingress endpoint while we mutate a Virtual Service (indirectly through Knative's internal type).

The type of mutations that occur include adding header responses and changing the backend destination. The backend destinations are static and always available.

This test has been flakey in earlier versions (https://github.com/knative-sandbox/net-istio/issues/938) but trying to bump to Istio 1.17.1 seems to consistently fail (https://github.com/knative-sandbox/net-istio/pull/1068)

### Version

```prose
Istio 1.14+ - more flakey with 1.16/1.17
```

**Comment by dprotaso** [comment]
Currently we have Istio 1.17.0 in CI - and you can see the flakey behaviour - https://testgrid.k8s.io/r/knative-own-testgrid/net-istio#continuous&width=20

**Comment by dprotaso** [comment]
Besides access logs are there ways to capture data or debug this further?

This might be similar to a bug we've seen in Contour - https://github.com/projectcontour/contour/issues/2151

**Comment by howardjohn** [comment]
Do we have access logs? That would eb the most useful to figure out what type of error we are talking about. Knowing what type of mutation would help as well. I couldn't reproduce just movbing a VS from backend1 to backend2

**Comment by dprotaso** [comment]
Test is here

https://github.com/knative/networking/blob/c692e9e6afe17fdd870e19fd04a869585296e315/test/conformance/ingress/update.go#L35

We actually stress it by running it 10x in parallel

https://github.com/knative-sandbox/net-istio/blob/11ffd8009173b43a2401a37245dc281806b6273c/test/conformance/ingress_test.go#L32

**Comment by howardjohn** [comment]
I tried to reproduce but it just hanged forever, this is after installing istio + knative and running `go test -count=1 -p 1 -tags=e2e ./test/conformance -run TestIngressConformance -v`.

**Comment by dprotaso** [comment]
> I tried to reproduce but it just hanged forever, this is after installing istio + knative and running go test -count=1 -p 1 -tags=e2e ./test/conformance -run TestIngressConformance -v.

There are some prereqs - install `ko` then

```sh
# Upload test images
$ ./test/upload-test-images.sh

# install our Istio setup
$ kubectl apply -f third_party/istio-latest/istio-ci-no-mesh/istio.yaml

# Install test resources
$ ko apply -f test/config 

# Install knative net-istio plugin
$ ko apply -f config/

# Run tests
$ gotestsum -f testname --  -tags e2e -timeout 30m --parallel 12 -run TestIngressConformance ./test/conformance --enable-beta --enable-alpha
```

### Failed run with Telemetry

I have a failed run. You can access [prow logs](https://prow.knative.dev/view/gs/knative-prow/pr-logs/pull/knative-sandbox_net-istio/1080/integration-tests_net-istio_main/1634252237339365376). Or it might be easier to access [Cluster Logs via GCP LogViewer](https://console.cloud.google.com/logs/query;query=resource.labels.cluster_name%3D%22e2e-cls-zpyvoajp%22;timeRange=2023-03-10T17:48:00.000Z%2F2023-03-10T18:39:00.000Z;cursorTimestamp=2023-03-10T18:28:55.470682659Z?referrer=search&project=knative-boskos-90)

You can access those logs by being a member of the google group https://groups.google.com/g/knative-dev


Here's a [query](https://console.cloud.google.com/logs/query;query=resource.labels.cluster_name%3D%22e2e-cls-zpyvoajp%22%0A-resource.labels.container_name%3D%22foo%22%0A%22ingress-conformance-0-update-tnemuddb%22%0A;timeRange=2023-03-10T18:06:30.000Z%2F2023-03-10T18:13:30.000Z;cursorTimestamp=2023-03-10T18:11:19.730387Z;resultsSearch=503?referrer=search&project=knative-boskos-90) that shows a pod becoming ready and then Istio serving 503s

You can see the Endpoint is ready around 2023-03-10 13:11:20.742 EST, insert ID(f53228af-5116-4f45-852a-13eaa9d68320)

> [2023-03-10T18:11:20.516Z] "GET / HTTP/1.1" 503 UH no_healthy_upstream - "-" 0 19 0 - "10.120.2.1" "knative.dev/TestIngressConformance/0/update/ingress-conformance-0-update-lmjbdirz" "31e6033c-8ad8-4989-9996-544bc574738f" "ingress-conformance-0-update-yaljemdf.example.com" "-" outbound|81||ingress-conformance-0-update-tnemuddb.serving-tests.svc.cluster.local - 10.120.2.9:8080 10.120.2.1:55564 - -

**Comment by dprotaso** [comment]
Here's a summary of the customizations we do

from: https://github.com/knative-sandbox/net-istio/blob/main/third_party/istio-latest/istio-ci-no-mesh.yaml

```yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  values:
    pilot:
      autoscaleMin: 3
      autoscaleMax: 10
      cpu:
        targetAverageUtilization: 60
    gateways:
      istio-ingressgateway:
        autoscaleEnabled: false

  meshConfig:
    defaultConfig:
      terminationDrainDuration: "20s"

  components:
    ingressGateways:
      - name: istio-ingressgateway
        k8s:
          replicaCount: 3
          resources:
            limits:
              cpu: 3000m
              memory: 2048Mi
            requests:
              cpu: 1000m
              memory: 1024Mi
```

**Comment by howardjohn** [comment]
Ok so in the logs of the prow failure.

I see ingress-conformance-0-update-tnemuddb is marked as "healthy" in endpoints. Then I see /healthz calls to it giving 503 UH. So far seems expected, there is eventual consistency to readiness. Then there are 503 calls to `/` - this seems unexpected maybe, I wouldn't think we move traffic over while the prober is failing(?). Eventually we get 200s.

This seems like expected behavior from Istio. There is eventual consistency, so we expect some 503s before we give 200s.

I have an extremely hard time following Knative tests, I've looked at the logs for like 30min and cannot understand what they are doing, what they are expecting, and what the outcome is. So all I understand is that we get 503s then some 200s which is in line with endpoint propagation. My expectation is knative is probing for 200s before switching traffic over, and the logs seem to indicate that is NOT happening.

In an "Istio bug" I would expect traffic to be:
* Probe 503... repeat as much as needed
* Probe 200
* Real traffic 503 - should be ready at this point

But we see, as far as I can tell:
* Probe 503... repeat as much as needed
* Real traffic 503 - ?? shouldn't be sending real traffic
* Probe 200
* Real traffic 200

I am probably missing something, let me know

**Comment by dprotaso** [comment]
Looks like this issue was reported before - https://github.com/istio/istio/issues/29310

But generally the test does the following
1. We have a VS pointing to a ready backend A (Deployment & Service)
1. There's prober running hitting the VS host name
1. We spin up backend B (Deployment & Service) and wait for a ready endpoint
1. We update the VS with the new backend B.
1. Check the prober only received 200 OKs

So the test is ensuring that updating a VS to point to a different and healthy backend is a safe operation. But we're in fact getting 503s.

These logs show the backend spinning up and becoming ready along with envoy cds logs - scroll to the beginning

https://console.cloud.google.com/logs/query;query=resource.labels.cluster_name%3D%22e2e-cls-vdqcepgz%22%0Atimestamp%3C%3D%222023-03-19T21:06:53.877850059Z%22%0Atimestamp%3E%3D%222023-03-19T21:00:00Z%22%0A%22ingress-conformance-9-update-xkeuftpr%22;timeRange=2023-03-19T21:06:52.816502Z%2F;summaryFields=resource%252Flabels%252Fpod_name:false:32:beginning;cursorTimestamp=2023-03-19T21:06:50Z;resultsSearch=ingress-conformance-9-update-xkeuftpr?referrer=search&project=knative-boskos-36

time | log line
-|-
2023-03-19T21:06:52.565402Z | Endpoint appears as ready  
2023-03-19T21:06:52.773955Z |	info	ads	Push debounce stable[1609] 6 for config ServiceEntry/serving-tests/ingress-conformance-9-update-xkeuftpr.serving-tests.svc.cluster.local and 3 more configs: 100.480777ms since last change, 242.27403ms since last push, full=true
.  | push debounce occurs 3x
2023-03-19T21:06:52.816502Z | Virtual Service is updated
2023-03-19T21:06:52.833554Z |	debug	envoy upstream external/envoy/source/common/upstream/cds_api_helper.cc:55	cds: add/update cluster 'outbound|72||ingress-conformance-9-update-xkeuftpr.serving-tests.svc.cluster.local' skipped	thread=16
. | same as above x5
2023-03-19T21:06:53.015959Z	| 503s - debug	envoy router external/envoy/source/common/router/router.cc:470	[C4467][S1965534937268489447] cluster 'outbound|72||ingress-conformance-9-update-xkeuftpr.serving-tests.svc.cluster.local' match for URL '/'	thread=26

**Comment by howardjohn** [comment]
That test seems invalid on any version of Istio and probably even stock kube-proxy. The workload being ready is the *signal* to the data plane to start sending traffic to the backend, not the data plane reporting it has actuated this. So there is expected eventual consistency here I think? lmk if I misunderstood

**Comment by howardjohn** [comment]
although the fact in that example it does seem like we did update the proxy is odd. I guess the eventual consistency is mitigated by the fact the updates to Ready and Virtual Service are through the same process (Istiod) so it *probably* typically processes them in that order (or together with debounce). However this is not guaranteed by k8s - Istio could see the virtual service event before the endpoint event I believe

**Comment by dprotaso** [comment]
> This test has been flakey in earlier versions (https://github.com/knative-sandbox/net-istio/issues/938)

It's been a problem in the past the issue above was filed when we were using Istio 1.13/4. We've been dealing this flakes for a while - I just stepped up to dig into their cause recently.

> The workload being ready is the signal to the data plane to start sending traffic to the backend, not the data plane reporting it has actuated this. So there is expected eventual consistency here I think? lmk if I misunderstood

But when the workload is ready it's currently not a backend for any VS. So at that time traffic is still going to backend A.

> So there is expected eventual consistency here I think? lmk if I misunderstood

I think I understand what you're saying - there's a race between the endpoint propagation and the route propagation.

But the drift seems pretty significant. Endpoint propagation is much slower and the first non-503 takes ~900ms from when the endpoint is ready. While the route cut over happens in 450ms (time from VS update to first 503)

**Comment by howardjohn** [comment]
Ok, got some more info.

so https://github.com/istio/istio/issues/29310 was a bug because we tried to remove un-unused envoy clusters. So when the route changed to point to them, the new cluster was added, but was still "warming" when the route was ready. This was reverted to the old behavior to fix the problem; the cluster is always there ready to go.

However, this still has an issue - when a cluster _changes_ in Envoy, it is basically a new cluster. When we scale from 0 to 1 instances, we are always going to change the cluster since we change the match_subject_alt_names. This introduces the same issue as 29310, but only scoped to "scale 0 to 1 then try to send traffic immediately"
